SIPS/sip-30.md

```diff
@@ -0,0 +1,238 @@
+"NOT RECOMMENDED", "MAY", and "OPTIONAL" written in uppercase in this document are to be interpreted as described in
+[RFC 2119](https://www.ietf.org/rfc/rfc2119.txt)
+
+### Snap Manifest
+
+A new set permission is added to the snap manifest:
+
+```json
+{
+  "initialPermissions": {
+    "snap_listAvailableEntropySources": {}
+  }
+}
+```
+
+
+### Common Types
+```typescript
+type Slip10Node = {
+  depth: number;
+  parentFingerprint: number;
+  index: number;
+  privateKey: string;
+  publicKey: string;
+  chainCode: string;
+  curve: "ed25519" | "secp256k1";
+};
+
+export type BIP44Node = {
+  coin_type: number;
+  depth: number;
+  privateKey: string;
+  publicKey: string;
+  chainCode: string;
+  path: string[];
+};
+
+interface EntropySource {
+  name: string;
+  id: string;
+  type: "mnemonic";
+}
+```
+
+### Scope
+
+This SIP applies to snaps that implement the [Keyring API][keyring-api] and any others which use the `snap_getEntropy`, `snap_getBip44Entropy`, `snap_getBip32Entropy`, and `snap_getBip32PublicKey` methods.
+
+### Snap Manifest
+
+No changes are required to the snap manifest.
+
+### Client wallet implementation
+
+If a snap requests a list of available entropy sources, and it has the permission to do so, the wallet MUST return a list of `EntropySource` objects.
+
+If a snap requests entropy and includes the `source` parameter for an entropy source of type `mnemonic`, the wallet MUST return entropy corresponding to that source, if it exists.
+
+If the source does not exist, the wallet MUST respond with an error.
+
+If the request does not include the `source` parameter, the wallet MUST return entropy from the default source.
+
+A client wallet MAY invoke the `keyring.createAccount` method with an `entropySource` parameter in the `options` object.
+
+The `entropySource` parameter MUST be a string which uniquely identifies the entropy source to use. It is not guaranteed to be the same string visible to any other snap, but should always refer to the same source in the context of interactions between the snap and the client wallet.
+
+#### Snap implementation
+
+If a snap is asked to create an account via `keyring.createAccount`, and the `entropySource` parameter is provided, and the snap requires entropy to create an account,the snap SHOULD request the entropy from the specified source.
+
+### New RPC Methods
+
+#### `snap_listAvailableEntropySources`
+
+The method returns an array of `EntropySource` objects, each representing an available entropy source. It is intended that the snap will display this list to the user.
+
+```typescript
+const entropySources = await snap.request({
+  method: "snap_listAvailableEntropySources",
+});
+// [
+//   { name: "Phrase 1", id: "phrase-1" },
+//   { name: "Phrase 2", id: "phrase-2" },
+// ]
+```
+
+### Existing RPC Methods
+
+#### `snap_getEntropy`
+
+##### Parameters
+An object containing:
+
+- `version` - The number 2. Version `2` is the first version that allows specifying a source ID.
```

Comment:
We don't need to increment the versioning number since the `source` parameter is optional and nothing else needs to change for the entropy derivation (except the source)